### PR TITLE
feat: implemented initial version of schema validation #195

### LINKED ISSUE #195: Schema validation
### Discussed in https://github.com/testinggospels/camouflage/discussions/194

<div type='discussions-op-text'>

<sup>Originally posted by **richardruiter** October 23, 2022</sup>
Hi, we are implementing Contract Testing and have a need for validating the mocks we create against OpenApi and XML-RPC schemas from the backend teams. This makes sure we don't make false assumptions when using a mock server.

We already have a homegrown mockserver with this principle baked in, but we need more functionality and camouflage covers it.
First I was thinking to create middleware that can be shared with multiple teams and created a successful PoC with `middleware.js`

However then the idea arose to make schema validation part of camouflage server.

```
loglevel: info
protocols:
  http:
    enable: true
    mocks_dir: "./mocks"
    port: 8080
validation:
  schemas:
    - https://raw.githubusercontent.com/OAI/OpenAPI-Specification/main/examples/v3.0/petstore.json
    - https://raw.githubusercontent.com/OAI/OpenAPI-Specification/main/examples/v3.0/example2.json
```

In the case of OpenApi It would parse the schemas with `@apidevtools/swagger-parser`, validate the request with `openapi-request-validator` and response with `openapi-response-validator.
</div>

### PR DESCRIPTION
**Pull request by richardruiter**
# Description

Please include a summary of the change and which issue is fixed. 

Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes #195

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

**Test Configuration**:
* OS: OSX
* Node version: 18.4.0
* NPM Version: 8.12.1

# Checklist:

- [x] My code follows the guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] This does not break any existing functionalities
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by richardruiter** [comment]
I made changes to the HttpParser that I'm insecure about. I needed to postpone sending a status and/or body in order to validate it. Need guidance to see if I don't break anything.

**Comment by shubhendumadhukar** [comment]
Changes look good. Some modification suggestions in HttpParser:
* Redundant reassignments of status code

https://github.com/richardruiter/camouflage/blob/082473b24c3fd8e60ef034bae6a56a7e8ea9f717/src/parser/HttpParser.ts#L91

https://github.com/richardruiter/camouflage/blob/082473b24c3fd8e60ef034bae6a56a7e8ea9f717/src/parser/HttpParser.ts#L169

https://github.com/richardruiter/camouflage/blob/082473b24c3fd8e60ef034bae6a56a7e8ea9f717/src/parser/HttpParser.ts#L187

First two instances (Line 91 and 169 can be removed), whereas 3rd instance can be modified from `response.status = codeResponse["status"] || this.res.statusCode;` to `response.status = codeResponse["status"];`

* Finally we should append headers to the response object, incase validations need to carried out on headers as well. We can add `response.headers[headerKey] = headerValue;` after following line

https://github.com/richardruiter/camouflage/blob/082473b24c3fd8e60ef034bae6a56a7e8ea9f717/src/parser/HttpParser.ts#L160

File: src/parser/HttpParser.ts [hunk 1]
```diff
@@ -64,20 +70,26 @@ export class HttpParser {
...
-        this.res.statusCode = response.status;
-        this.res.set(response.headers)
-        this.res.send(response.body);
+        logger.debug(
+          `No custom global override for default response. Sending default Camouflage response.`
+        );
       }
     }
+
+    response.status = parseFloat((<unknown>response.status) as string);
```
Comment (richardruiter) [inline src/parser/HttpParser.ts:91]: This converts the status code from string to number
```
export interface HttpParserResponse {
  status: number;
```

So in that sense it's not redundant.

Comment (shubhendumadhukar) [reply src/parser/HttpParser.ts:91]: I think we have covered all the assignments:
1. The response has been initialised  with `status: 404,` a number.
2. Any assignments to `response.status` in `prepareResponse` method, takes care of the conversion from string to number. `response.status = <number>(<unknown>line.match(regex)[1]);`

Apart from these two cases, I am not sure there would be any other use cases where the conversion at L91 makes sense, however we could leave it there to catch any unhandled scenarios (e.g. L187 `response.status = codeResponse["status"];`)

Comment (richardruiter) [reply src/parser/HttpParser.ts:91]: 👍

**Comment by richardruiter** [comment]
Added docs. Can we merge it?

**Pull request merged**